File: worker/tasks/recurring.go
package tasks

import (
	"context"
	"fmt"
	"log"
	"time"

	"retrospend-worker/db"
)

type RecurringTemplate struct {
	ID          string
	UserID      string
	Name        string
	Amount      float64
	Currency    string
	CategoryID  *string
	Frequency   string
	NextDueDate time.Time
}

func calculateNextDueDate(current time.Time, frequency string) time.Time {
	next := current
	switch frequency {
	case "WEEKLY":
		next = next.AddDate(0, 0, 7)
	case "MONTHLY":
		next = next.AddDate(0, 1, 0)
	case "YEARLY":
		next = next.AddDate(1, 0, 0)
	}
	return next
}

func getBestExchangeRate(ctx context.Context, database *db.DB, currency string, date time.Time) (float64, error) {
	if currency == "USD" {
		return 1.0, nil
	}

	var rate float64
	err := database.Pool.QueryRow(ctx, `
		SELECT rate FROM exchange_rate 
		WHERE currency = $1 AND date <= $2
		ORDER BY 
			CASE WHEN type = 'blue' THEN 0 WHEN type = 'official' THEN 1 ELSE 2 END,
			date DESC
		LIMIT 1
	`, currency, date).Scan(&rate)

	if err != nil {
		return 0, fmt.Errorf("no exchange rate found for %s: %w", currency, err)
	}

	return rate, nil
}

func ProcessRecurringExpenses(database *db.DB) error {
	log.Println("[RECURRING] Processing due expenses...")

	ctx := context.Background()
	now := time.Now()

	// Get all due templates
	// Note: using camelCase "userId", "categoryId", "nextDueDate", "isActive", "autoPay"
	rows, err := database.Pool.Query(ctx, `
		SELECT id, "userId", name, amount, currency, "categoryId", frequency, "nextDueDate"
		FROM recurring_template
		WHERE "isActive" = true 
		  AND "autoPay" = true 
		  AND "nextDueDate" <= $1
	`, now)
	if err != nil {
		return fmt.Errorf("failed to query templates: %w", err)
	}
	defer rows.Close()

	var templates []RecurringTemplate
	for rows.Next() {
		var t RecurringTemplate
		if err := rows.Scan(&t.ID, &t.UserID, &t.Name, &t.Amount, &t.Currency, &t.CategoryID, &t.Frequency, &t.NextDueDate); err != nil {
			return fmt.Errorf("failed to scan template: %w", err)
		}
		templates = append(templates, t)
	}

	if len(templates) == 0 {
		log.Println("[RECURRING] No due expenses found")
		// Still update status to show we checked
		if err := database.UpdateWorkerStatus(ctx, "recurring_expenses", true); err != nil {
			log.Printf("[RECURRING] Failed to update worker status: %v", err)
		}
		return nil
	}

	createdCount := 0

	for _, template := range templates {
		// Get exchange rate
		exchangeRate, err := getBestExchangeRate(ctx, database, template.Currency, template.NextDueDate)
		if err != nil {
			log.Printf("[RECURRING] Warning: %v, skipping template %s", err, template.ID)
			continue
		}

		amountInUSD := template.Amount / exchangeRate

		// Create expense
		// Note: "userId", "categoryId", "amountInUSD", "exchangeRate", "pricingSource", "recurringTemplateId", etc
		_, err = database.Pool.Exec(ctx, `
			INSERT INTO expense (
				id, "userId", title, amount, currency, date, "categoryId",
				"amountInUSD", "exchangeRate", "pricingSource", "recurringTemplateId", 
				status, "createdAt", "updatedAt"
			) VALUES (
				gen_random_uuid()::text, $1, $2, $3, $4, $5, $6, $7, $8, 'RECURRING', $9, 'FINALIZED', NOW(), NOW()
			)
		`, template.UserID, template.Name, template.Amount, template.Currency, template.NextDueDate,
			template.CategoryID, amountInUSD, exchangeRate, template.ID)

		if err != nil {
			log.Printf("[RECURRING] Error creating expense for template %s: %v", template.ID, err)
			continue
		}

		// Update next due date
		nextDueDate := calculateNextDueDate(template.NextDueDate, template.Frequency)
		_, err = database.Pool.Exec(ctx, `
			UPDATE recurring_template 
			SET "nextDueDate" = $1, "updatedAt" = NOW() 
			WHERE id = $2
		`, nextDueDate, template.ID)

		if err != nil {
			log.Printf("[RECURRING] Error updating template %s: %v", template.ID, err)
			continue
		}

		createdCount++
	}
	
	log.Printf("[RECURRING] ✓ Processed %d templates, created %d expenses", len(templates), createdCount)
	
	// Update worker status
	if err := database.UpdateWorkerStatus(ctx, "recurring_expenses", true); err != nil {
		log.Printf("[RECURRING] Failed to update worker status: %v", err)
	}

	return nil
}

File: src/server/api/routers/recurring.ts
import { TRPCError } from "@trpc/server";
import { z } from "zod";
import { BASE_CURRENCY } from "~/lib/constants";
import { createTRPCRouter, protectedProcedure } from "~/server/api/trpc";
import { fromUSD, toUSD } from "~/server/currency";
import { RateCache } from "~/server/services/rate-cache";
import { getBestExchangeRate } from "./shared-currency";

// Helper to calculate next due date based on frequency
function calculateNextDueDate(
	currentDate: Date,
	frequency: "WEEKLY" | "BIWEEKLY" | "MONTHLY" | "QUARTERLY" | "YEARLY",
): Date {
	const next = new Date(currentDate);

	switch (frequency) {
		case "WEEKLY":
			next.setDate(next.getDate() + 7);
			break;
		case "BIWEEKLY":
			next.setDate(next.getDate() + 14);
			break;
		case "MONTHLY":
			next.setMonth(next.getMonth() + 1);
			break;
		case "QUARTERLY":
			next.setMonth(next.getMonth() + 3);
			break;
		case "YEARLY":
			next.setFullYear(next.getFullYear() + 1);
			break;
	}

	return next;
}

export const recurringRouter = createTRPCRouter({
	list: protectedProcedure
		.input(
			z
				.object({
					includeInactive: z.boolean().optional(),
				})
				.optional(),
		)
		.query(async ({ ctx, input }) => {
			const { session, db } = ctx;

			const templates = await db.recurringTemplate.findMany({
				where: {
					userId: session.user.id,
					...(input?.includeInactive ? {} : { isActive: true }),
				},
				orderBy: {
					nextDueDate: "asc",
				},
				include: {
					category: {
						select: {
							id: true,
							name: true,
							color: true,
						},
					},
				},
			});

			const user = await db.user.findUnique({
				where: { id: session.user.id },
				select: { homeCurrency: true },
			});
			const homeCurrency = user?.homeCurrency ?? "USD";

			// Batch-fetch all needed exchange rates in one pass
			const currenciesNeeded = new Set<string>();
			for (const t of templates) {
				if (t.currency !== homeCurrency) {
					currenciesNeeded.add(t.currency);
					if (homeCurrency !== "USD") currenciesNeeded.add(homeCurrency);
				}
			}

			const rateCache = new RateCache(db);
			const rateMap = await rateCache.getMany([...currenciesNeeded], new Date());

			const enrichedTemplates = templates.map((t) => {
				let amountInHomeCurrency = Number(t.amount);
				let exchangeRate = 1;

				if (t.currency !== homeCurrency) {
					const bestRateFrom = rateMap.get(t.currency);

					let usdValue = Number(t.amount);
					if (t.currency !== "USD" && bestRateFrom) {
						exchangeRate = bestRateFrom.rate;
						usdValue = toUSD(Number(t.amount), t.currency, bestRateFrom.rate);
					}

					if (homeCurrency === "USD") {
						amountInHomeCurrency = usdValue;
					} else {
						const bestRateTo = rateMap.get(homeCurrency);
						if (bestRateTo) {
							amountInHomeCurrency = fromUSD(usdValue, homeCurrency, bestRateTo.rate);
						} else {
							amountInHomeCurrency = usdValue;
						}
					}
				}

				return {
					...t,
					amountInHomeCurrency,
					exchangeRate,
				};
			});

			return enrichedTemplates;
		}),

	get: protectedProcedure
		.input(
			z.object({
				id: z.string().cuid(),
			}),
		)
		.query(async ({ ctx, input }) => {
			const { session, db } = ctx;

			const template = await db.recurringTemplate.findFirst({
				where: {
					id: input.id,
					userId: session.user.id,
				},
				include: {
					category: {
						select: {
							id: true,
							name: true,
							color: true,
						},
					},
					expenses: {
						select: {
							id: true,
							title: true,
							amount: true,
							date: true,
						},
						orderBy: {
							date: "desc",
						},
						take: 10,
					},
				},
			});

			if (!template) {
				throw new TRPCError({
					code: "NOT_FOUND",
					message: "Recurring template not found",
				});
			}

			return template;
		}),

	create: protectedProcedure
		.input(
			z.object({
				name: z.string().min(1).max(191),
				amount: z.number().positive(),
				currency: z.string().length(3).default("USD"),
				categoryId: z.string().cuid().optional(),
				frequency: z.enum(["WEEKLY", "BIWEEKLY", "MONTHLY", "QUARTERLY", "YEARLY"]),
				nextDueDate: z.date(),
				websiteUrl: z.string().url().max(512).optional(),
				paymentSource: z.string().min(1).max(191).optional(),
				autoPay: z.boolean().default(true),
			}),
		)
		.mutation(async ({ ctx, input }) => {
			const { session, db } = ctx;

			// Validate category if provided
			if (input.categoryId) {
				const category = await db.category.findFirst({
					where: { id: input.categoryId, userId: session.user.id },
				});
				if (!category) {
					throw new TRPCError({
						code: "FORBIDDEN",
						message: "Category not found",
					});
				}
			}

			const template = await db.recurringTemplate.create({
				data: {
					userId: session.user.id,
					name: input.name,
					amount: input.amount,
					currency: input.currency,
					categoryId: input.categoryId,
					frequency: input.frequency,
					nextDueDate: input.nextDueDate,
					websiteUrl: input.websiteUrl,
					paymentSource: input.paymentSource,
					autoPay: input.autoPay,
				},
				include: {
					category: {
						select: {
							id: true,
							name: true,
							color: true,
						},
					},
				},
			});

			return template;
		}),

	update: protectedProcedure
		.input(
			z.object({
				id: z.string().cuid(),
				name: z.string().min(1).max(191).optional(),
				amount: z.number().positive().optional(),
				currency: z.string().length(3).optional(),
				categoryId: z.string().cuid().nullable().optional(),
				frequency: z.enum(["WEEKLY", "BIWEEKLY", "MONTHLY", "QUARTERLY", "YEARLY"]).optional(),
				nextDueDate: z.date().optional(),
				websiteUrl: z.string().url().max(512).nullable().optional(),
				paymentSource: z.string().min(1).max(191).nullable().optional(),
				autoPay: z.boolean().optional(),
				isActive: z.boolean().optional(),
			}),
		)
		.mutation(async ({ ctx, input }) => {
			const { session, db } = ctx;

			const existing = await db.recurringTemplate.findFirst({
				where: {
					id: input.id,
					userId: session.user.id,
				},
			});

			if (!existing) {
				throw new TRPCError({
					code: "NOT_FOUND",
					message: "Recurring template not found",
				});
			}

			// Validate category if provided
			if (input.categoryId) {
				const category = await db.category.findFirst({
					where: { id: input.categoryId, userId: session.user.id },
				});
				if (!category) {
					throw new TRPCError({
						code: "FORBIDDEN",
						message: "Category not found",
					});
				}
			}

			const template = await db.recurringTemplate.update({
				where: { id: input.id },
				data: {
					name: input.name,
					amount: input.amount,
					currency: input.currency,
					categoryId: input.categoryId,
					frequency: input.frequency,
					nextDueDate: input.nextDueDate,
					websiteUrl: input.websiteUrl,
					paymentSource: input.paymentSource,
					autoPay: input.autoPay,
					isActive: input.isActive,
				},
				include: {
					category: {
						select: {
							id: true,
							name: true,
							color: true,
						},
					},
				},
			});

			return template;
		}),

	delete: protectedProcedure
		.input(
			z.object({
				id: z.string().cuid(),
			}),
		)
		.mutation(async ({ ctx, input }) => {
			const { session, db } = ctx;

			const result = await db.recurringTemplate.deleteMany({
				where: {
					id: input.id,
					userId: session.user.id,
				},
			});

			if (result.count === 0) {
				throw new TRPCError({
					code: "NOT_FOUND",
					message: "Recurring template not found",
				});
			}

			return { success: true };
		}),

	/**
	 * Get templates that are pending confirmation (autoPay = false and due)
	 */
	getPendingConfirmation: protectedProcedure.query(async ({ ctx }) => {
		const { session, db } = ctx;
		const now = new Date();

		const pending = await db.recurringTemplate.findMany({
			where: {
				userId: session.user.id,
				isActive: true,
				autoPay: false,
				nextDueDate: {
					lte: now,
				},
			},
			include: {
				category: {
					select: {
						id: true,
						name: true,
						color: true,
					},
				},
			},
			orderBy: {
				nextDueDate: "asc",
			},
		});

		return pending;
	}),

	/**
	 * Confirm a pending recurring template and create the expense
	 */
	confirmAndCreate: protectedProcedure
		.input(
			z.object({
				id: z.string().cuid(),
			}),
		)
		.mutation(async ({ ctx, input }) => {
			const { session, db } = ctx;

			const template = await db.recurringTemplate.findFirst({
				where: {
					id: input.id,
					userId: session.user.id,
					isActive: true,
				},
			});

			if (!template) {
				throw new TRPCError({
					code: "NOT_FOUND",
					message: "Recurring template not found",
				});
			}

			// Get exchange rate
			let exchangeRate = 1;
			let amountInUSD = Number(template.amount);

			if (template.currency !== BASE_CURRENCY) {
				const bestRate = await getBestExchangeRate(
					db,
					template.currency,
					template.nextDueDate,
				);
				if (bestRate) {
					exchangeRate = bestRate.rate;
					amountInUSD = toUSD(Number(template.amount), template.currency, exchangeRate);
				}
			}

			// Calculate next due date
			const nextDueDate = calculateNextDueDate(
				template.nextDueDate,
				template.frequency,
			);

			// Create expense and update template in a transaction
			// This prevents orphaned expenses if the nextDueDate update fails
			const expense = await db.$transaction(async (tx) => {
				await tx.$executeRaw`SELECT set_config('app.current_user_id', ${session.user.id}, true),
				                            set_config('role', 'retrospend_app', true)`;
				const createdExpense = await tx.expense.create({
					data: {
						userId: session.user.id,
						title: template.name,
						amount: template.amount,
						currency: template.currency,
						date: template.nextDueDate,
						categoryId: template.categoryId,
						amountInUSD,
						exchangeRate,
						pricingSource: "RECURRING",
						recurringTemplateId: template.id,
						status: "FINALIZED",
					},
				});

				// Update next due date
				await tx.recurringTemplate.update({
					where: { id: template.id },
					data: { nextDueDate },
				});

				return createdExpense;
			});

			return expense;
		}),
});
